File: clients/python/upload_client.py
"""
TUS Python Client

A comprehensive TUS (Tus Resumable Upload Protocol) client implementation
with support for:
- Resumable uploads
- Progress tracking
- Metadata handling
- Error handling and retry logic
- Chunked uploads
- Parallel uploads
"""

import os
import sys
import time
import base64
import hashlib
import mimetypes
from typing import Dict, Any, Optional, Callable, Union
from pathlib import Path

import requests
from tqdm import tqdm

class TUSUploadError(Exception):
    """Exception raised for TUS upload errors."""
    pass

class TUSClient:
    """
    TUS (Tus Resumable Upload Protocol) client for Python.
    
    Provides a high-level interface for uploading files using the TUS protocol
    with support for resumable uploads, progress tracking, and metadata.
    """
    
    def __init__(self, 
                 endpoint: str,
                 chunk_size: int = 4 * 1024 * 1024,  # 4MB default
                 timeout: int = 30,
                 max_retries: int = 3,
                 retry_delay: float = 1.0):
        """
        Initialize TUS client.
        
        Args:
            endpoint: TUS server endpoint URL
            chunk_size: Size of each upload chunk in bytes
            timeout: Request timeout in seconds
            max_retries: Maximum number of retries for failed requests
            retry_delay: Delay between retries in seconds
        """
        self.endpoint = endpoint.rstrip('/')
        self.chunk_size = chunk_size
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        
        # Default headers
        self.headers = {
            'Tus-Resumable': '1.0.0',
            'User-Agent': 'TUS-Python-Client/1.0.0'
        }
        
        # Session for connection pooling
        self.session = requests.Session()
        self.session.headers.update(self.headers)
    
    def upload(self,
               file_path: Union[str, Path],
               metadata: Optional[Dict[str, str]] = None,
               progress_callback: Optional[Callable[[int, int], None]] = None,
               resume: bool = True) -> 'UploadResult':
        """
        Upload a file using TUS protocol.
        
        Args:
            file_path: Path to the file to upload
            metadata: Optional metadata dictionary
            progress_callback: Optional callback function for progress updates
            resume: Whether to attempt resuming interrupted uploads
            
        Returns:
            UploadResult object with upload information
            
        Raises:
            TUSUploadError: If upload fails
        """
        file_path = Path(file_path)
        
        if not file_path.exists():
            raise TUSUploadError(f"File not found: {file_path}")
        
        if not file_path.is_file():
            raise TUSUploadError(f"Not a file: {file_path}")
        
        file_size = file_path.stat().st_size
        
        if file_size == 0:
            raise TUSUploadError("Cannot upload empty file")
        
        # Prepare metadata
        if metadata is None:
            metadata = {}
        
        # Auto-detect filename and filetype if not provided
        if 'filename' not in metadata:
            metadata['filename'] = file_path.name
        
        if 'filetype' not in metadata:
            mime_type, _ = mimetypes.guess_type(str(file_path))
            if mime_type:
                metadata['filetype'] = mime_type
        
        # Add file size to metadata
        metadata['size'] = str(file_size)
        
        # Calculate file checksum for integrity verification
        file_checksum = self._calculate_checksum(file_path)
        metadata['sha256'] = file_checksum
        
        # Create upload
        upload_url = self._create_upload(file_size, metadata)
        
        # Upload file with progress tracking
        final_url = self._upload_file(file_path, upload_url, file_size, progress_callback)
        
        return UploadResult(
            url=final_url,
            size=file_size,
            metadata=metadata,
            checksum=file_checksum
        )
    
    def upload_bytes(self,
                    data: bytes,
                    filename: str,
                    metadata: Optional[Dict[str, str]] = None,
                    progress_callback: Optional[Callable[[int, int], None]] = None) -> 'UploadResult':
        """
        Upload bytes data using TUS protocol.
        
        Args:
            data: Bytes data to upload
            filename: Name for the uploaded file
            metadata: Optional metadata dictionary
            progress_callback: Optional callback function for progress updates
            
        Returns:
            UploadResult object with upload information
        """
        if not data:
            raise TUSUploadError("Cannot upload empty data")
        
        file_size = len(data)
        
        # Prepare metadata
        if metadata is None:
            metadata = {}
        
        metadata['filename'] = filename
        metadata['size'] = str(file_size)
        
        # Auto-detect filetype if not provided
        if 'filetype' not in metadata:
            mime_type, _ = mimetypes.guess_type(filename)
            if mime_type:
                metadata['filetype'] = mime_type
        
        # Calculate checksum
        file_checksum = hashlib.sha256(data).hexdigest()
        metadata['sha256'] = file_checksum
        
        # Create upload
        upload_url = self._create_upload(file_size, metadata)
        
        # Upload data with progress tracking
        final_url = self._upload_bytes(data, upload_url, file_size, progress_callback)
        
        return UploadResult(
            url=final_url,
            size=file_size,
            metadata=metadata,
            checksum=file_checksum
        )
    
    def resume_upload(self, upload_url: str, file_path: Union[str, Path],
                      progress_callback: Optional[Callable[[int, int], None]] = None) -> 'UploadResult':
        """
        Resume a previously interrupted upload.
        
        Args:
            upload_url: URL of the existing upload
            file_path: Path to the file being uploaded
            progress_callback: Optional callback function for progress updates
            
        Returns:
            UploadResult object with upload information
        """
        file_path = Path(file_path)
        file_size = file_path.stat().st_size
        
        # Get current offset
        current_offset = self._get_upload_offset(upload_url)
        
        if current_offset >= file_size:
            # Upload already completed
            return UploadResult(url=upload_url, size=file_size, metadata={}, checksum="")
        
        # Resume upload from current offset
        final_url = self._upload_file(file_path, upload_url, file_size, progress_callback, current_offset)
        
        return UploadResult(url=final_url, size=file_size, metadata={}, checksum="")
    
    def get_upload_info(self, upload_url: str) -> Dict[str, Any]:
        """
        Get information about an existing upload.
        
        Args:
            upload_url: URL of the upload
            
        Returns:
            Dictionary with upload information
        """
        try:
            response = self.session.head(upload_url, timeout=self.timeout)
            response.raise_for_status()
            
            return {
                'offset': int(response.headers.get('Upload-Offset', 0)),
                'length': int(response.headers.get('Upload-Length', 0)),
                'metadata': response.headers.get('Upload-Metadata', ''),
                'expires': response.headers.get('Upload-Expires'),
                'complete': response.headers.get('Upload-Offset') == response.headers.get('Upload-Length')
            }
        except requests.RequestException as e:
            raise TUSUploadError(f"Failed to get upload info: {e}")
    
    def delete_upload(self, upload_url: str) -> bool:
        """
        Delete an existing upload.
        
        Args:
            upload_url: URL of the upload to delete
            
        Returns:
            True if deletion was successful
        """
        try:
            response = self.session.delete(upload_url, timeout=self.timeout)
            response.raise_for_status()
            return True
        except requests.RequestException:
            return False
    
    def _create_upload(self, file_size: int, metadata: Dict[str, str]) -> str:
        """Create a new upload and return the upload URL."""
        headers = {
            'Upload-Length': str(file_size),
            'Content-Type': 'application/offset+octet-stream'
        }
        
        # Encode metadata
        if metadata:
            encoded_metadata = []
            for key, value in metadata.items():
                encoded_key = key
                encoded_value = base64.b64encode(value.encode('utf-8')).decode('ascii')
                encoded_metadata.append(f"{encoded_key} {encoded_value}")
            headers['Upload-Metadata'] = ','.join(encoded_metadata)
        
        try:
            response = self.session.post(self.endpoint, headers=headers, timeout=self.timeout)
            response.raise_for_status()
            
            upload_url = response.headers.get('Location')
            if not upload_url:
                raise TUSUploadError("Server did not return upload URL")
            
            # Handle relative URLs
            if upload_url.startswith('/'):
                from urllib.parse import urljoin
                upload_url = urljoin(self.endpoint, upload_url)
            
            return upload_url
            
        except requests.RequestException as e:
            raise TUSUploadError(f"Failed to create upload: {e}")
    
    def _upload_file(self, file_path: Path, upload_url: str, file_size: int,
                     progress_callback: Optional[Callable[[int, int], None]] = None,
                     start_offset: int = 0) -> str:
        """Upload file data to the given upload URL."""
        with open(file_path, 'rb') as file:
            file.seek(start_offset)
            return self._upload_stream(file, upload_url, file_size, progress_callback, start_offset)
    
    def _upload_bytes(self, data: bytes, upload_url: str, file_size: int,
                      progress_callback: Optional[Callable[[int, int], None]] = None,
                      start_offset: int = 0) -> str:
        """Upload bytes data to the given upload URL."""
        from io import BytesIO
        stream = BytesIO(data)
        stream.seek(start_offset)
        return self._upload_stream(stream, upload_url, file_size, progress_callback, start_offset)
    
    def _upload_stream(self, stream, upload_url: str, file_size: int,
                       progress_callback: Optional[Callable[[int, int], None]] = None,
                       start_offset: int = 0) -> str:
        """Upload stream data to the given upload URL."""
        current_offset = start_offset
        
        # Initialize progress bar if no callback provided
        if progress_callback is None:
            pbar = tqdm(
                total=file_size,
                initial=start_offset,
                unit='B',
                unit_scale=True,
                desc="Uploading"
            )
            progress_callback = lambda current, total: pbar.update(current - pbar.n)
        
        try:
            while current_offset < file_size:
                # Read chunk
                chunk = stream.read(min(self.chunk_size, file_size - current_offset))
                if not chunk:
                    break
                
                # Upload chunk with retry logic
                chunk_uploaded = False
                for attempt in range(self.max_retries):
                    try:
                        headers = {
                            'Upload-Offset': str(current_offset),
                            'Content-Type': 'application/offset+octet-stream'
                        }
                        
                        response = self.session.patch(
                            upload_url,
                            headers=headers,
                            data=chunk,
                            timeout=self.timeout
                        )
                        response.raise_for_status()
                        
                        # Update offset
                        new_offset = int(response.headers.get('Upload-Offset', current_offset))
                        current_offset = new_offset
                        
                        # Call progress callback
                        if progress_callback:
                            progress_callback(current_offset, file_size)
                        
                        chunk_uploaded = True
                        break
                        
                    except requests.RequestException as e:
                        if attempt < self.max_retries - 1:
                            time.sleep(self.retry_delay * (attempt + 1))
                            continue
                        else:
                            raise TUSUploadError(f"Failed to upload chunk after {self.max_retries} attempts: {e}")
                
                if not chunk_uploaded:
                    raise TUSUploadError("Failed to upload chunk")
            
            # Close progress bar if we created it
            if 'pbar' in locals():
                pbar.close()
            
            return upload_url
            
        except Exception as e:
            if 'pbar' in locals():
                pbar.close()
            raise
    
    def _get_upload_offset(self, upload_url: str) -> int:
        """Get the current upload offset."""
        try:
            response = self.session.head(upload_url, timeout=self.timeout)
            response.raise_for_status()
            return int(response.headers.get('Upload-Offset', 0))
        except requests.RequestException as e:
            raise TUSUploadError(f"Failed to get upload offset: {e}")
    
    def _calculate_checksum(self, file_path: Path) -> str:
        """Calculate SHA256 checksum of a file."""
        sha256_hash = hashlib.sha256()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                sha256_hash.update(chunk)
        return sha256_hash.hexdigest()


class UploadResult:
    """Result of a TUS upload operation."""
    
    def __init__(self, url: str, size: int, metadata: Dict[str, str], checksum: str):
        self.url = url
        self.size = size
        self.metadata = metadata
        self.checksum = checksum
        self.uploaded_at = time.time()
    
    def __str__(self):
        return f"UploadResult(url='{self.url}', size={self.size}, checksum='{self.checksum[:8]}...')"
    
    def __repr__(self):
        return self.__str__()


# Example usage
if __name__ == "__main__":
    # Example usage of the TUS client
    client = TUSClient("http://localhost:1080/files/")
    
    # Create a test file
    test_file = Path("/tmp/test_upload.txt")
    test_file.write_text("Hello, TUS! This is a test upload.\n" * 1000)
    
    try:
        print("Uploading test file...")
        result = client.upload(
            file_path=test_file,
            metadata={
                "description": "Test upload from Python client",
                "author": "TUS Python Client"
            },
            progress_callback=lambda current, total: print(f"Progress: {current/total*100:.1f}%")
        )
        
        print(f"Upload successful!")
        print(f"URL: {result.url}")
        print(f"Size: {result.size} bytes")
        print(f"Checksum: {result.checksum}")
        
    except TUSUploadError as e:
        print(f"Upload failed: {e}")
    finally:
        # Clean up test file
        if test_file.exists():
            test_file.unlink()

File: clients/golang/upload_client.go
package main

import (
	"context"
	"crypto/sha256"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/tus/tus-go-client"
)

// TUSClient wraps the tus-go-client with additional functionality
type TUSClient struct {
	client    *tus.Client
	endpoint  string
	chunkSize int64
	timeout   time.Duration
}

// UploadOptions defines options for upload operations
type UploadOptions struct {
	ChunkSize        int64
	Concurrency      int
	Metadata         map[string]string
	ProgressCallback func(current, total int64)
	Resume           bool
	Context          context.Context
}

// UploadResult contains the result of an upload operation
type UploadResult struct {
	URL      string
	Size     int64
	Metadata map[string]string
	Checksum string
	Duration time.Duration
}

// UploadProgress tracks upload progress with detailed statistics
type UploadProgress struct {
	current   int64
	total     int64
	startTime time.Time
	lastTime  time.Time
	lastBytes int64
	callback  func(current, total int64)
	mu        sync.RWMutex
}

// NewUploadProgress creates a new progress tracker
func NewUploadProgress(total int64, callback func(current, total int64)) *UploadProgress {
	return &UploadProgress{
		total:     total,
		startTime: time.Now(),
		lastTime:  time.Now(),
		callback:  callback,
	}
}

// Update updates the progress and calls the callback
func (p *UploadProgress) Update(current int64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	
	p.current = current
	now := time.Now()
	
	if p.callback != nil && (now.Sub(p.lastTime) > 100*time.Millisecond || current == p.total) {
		p.callback(current, p.total)
		p.lastTime = now
	}
}

// GetStats returns current progress statistics
func (p *UploadProgress) GetStats() (percent float64, speed float64, eta time.Duration) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	
	if p.total == 0 {
		return 0, 0, 0
	}
	
	percent = float64(p.current) / float64(p.total) * 100
	
	elapsed := time.Since(p.startTime)
	if elapsed.Seconds() > 0 && p.current > 0 {
		speed = float64(p.current) / elapsed.Seconds() // bytes per second
		remaining := p.total - p.current
		eta = time.Duration(float64(remaining)/speed) * time.Second
	}
	
	return percent, speed, eta
}

// NewTUSClient creates a new TUS client
func NewTUSClient(endpoint string) (*TUSClient, error) {
	tusConfig := tus.DefaultConfig()
	tusConfig.Resume = true
	tusConfig.HttpClient = &http.Client{
		Timeout: 30 * time.Second,
	}
	
	client, err := tus.NewClient(endpoint, tusConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create TUS client: %w", err)
	}
	
	return &TUSClient{
		client:    client,
		endpoint:  endpoint,
		chunkSize: 4 * 1024 * 1024, // 4MB default
		timeout:   30 * time.Second,
	}, nil
}

// UploadFile uploads a file with the given options
func (c *TUSClient) UploadFile(ctx context.Context, filePath string, options UploadOptions) (*UploadResult, error) {
	startTime := time.Now()
	
	// Open file
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()
	
	// Get file info
	fileInfo, err := file.Stat()
	if err != nil {
		return nil, fmt.Errorf("failed to get file info: %w", err)
	}
	
	fileSize := fileInfo.Size()
	filename := filepath.Base(filePath)
	
	// Calculate checksum
	checksum, err := calculateFileChecksum(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to calculate checksum: %w", err)
	}
	
	// Prepare metadata
	metadata := make(map[string]string)
	if options.Metadata != nil {
		for k, v := range options.Metadata {
			metadata[k] = v
		}
	}
	metadata["filename"] = filename
	metadata["size"] = strconv.FormatInt(fileSize, 10)
	metadata["sha256"] = checksum
	
	// Setup progress tracking
	var progress *UploadProgress
	if options.ProgressCallback != nil {
		progress = NewUploadProgress(fileSize, options.ProgressCallback)
	}
	
	// Create upload
	uploader, err := c.client.CreateUpload(&tus.Upload{
		Stream:   file,
		Size:     fileSize,
		Metadata: metadata,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create upload: %w", err)
	}
	
	// Configure uploader
	if options.ChunkSize > 0 {
		uploader.ChunkSize = options.ChunkSize
	} else {
		uploader.ChunkSize = c.chunkSize
	}
	
	// Setup progress callback for the uploader
	if progress != nil {
		uploader.NotifyUploadProgress = func(bytesUploaded, bytesTotal int64) {
			progress.Update(bytesUploaded)
		}
	}
	
	// Use context if provided
	if options.Context != nil {
		ctx = options.Context
	}
	if ctx == nil {
		ctx = context.Background()
	}
	
	// Perform upload
	err = uploader.Upload(ctx)
	if err != nil {
		return nil, fmt.Errorf("upload failed: %w", err)
	}
	
	// Final progress update
	if progress != nil {
		progress.Update(fileSize)
	}
	
	duration := time.Since(startTime)
	
	return &UploadResult{
		URL:      uploader.Url(),
		Size:     fileSize,
		Metadata: metadata,
		Checksum: checksum,
		Duration: duration,
	}, nil
}

// UploadBytes uploads byte data with the given options
func (c *TUSClient) UploadBytes(ctx context.Context, data []byte, filename string, options UploadOptions) (*UploadResult, error) {
	startTime := time.Now()
	
	if len(data) == 0 {
		return nil, fmt.Errorf("cannot upload empty data")
	}
	
	dataSize := int64(len(data))
	
	// Calculate checksum
	hash := sha256.Sum256(data)
	checksum := fmt.Sprintf("%x", hash)
	
	// Prepare metadata
	metadata := make(map[string]string)
	if options.Metadata != nil {
		for k, v := range options.Metadata {
			metadata[k] = v
		}
	}
	metadata["filename"] = filename
	metadata["size"] = strconv.FormatInt(dataSize, 10)
	metadata["sha256"] = checksum
	
	// Setup progress tracking
	var progress *UploadProgress
	if options.ProgressCallback != nil {
		progress = NewUploadProgress(dataSize, options.ProgressCallback)
	}
	
	// Create reader from bytes
	reader := strings.NewReader(string(data))
	
	// Create upload
	uploader, err := c.client.CreateUpload(&tus.Upload{
		Stream:   reader,
		Size:     dataSize,
		Metadata: metadata,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create upload: %w", err)
	}
	
	// Configure uploader
	if options.ChunkSize > 0 {
		uploader.ChunkSize = options.ChunkSize
	} else {
		uploader.ChunkSize = c.chunkSize
	}
	
	// Setup progress callback
	if progress != nil {
		uploader.NotifyUploadProgress = func(bytesUploaded, bytesTotal int64) {
			progress.Update(bytesUploaded)
		}
	}
	
	// Use context if provided
	if options.Context != nil {
		ctx = options.Context
	}
	if ctx == nil {
		ctx = context.Background()
	}
	
	// Perform upload
	err = uploader.Upload(ctx)
	if err != nil {
		return nil, fmt.Errorf("upload failed: %w", err)
	}
	
	// Final progress update
	if progress != nil {
		progress.Update(dataSize)
	}
	
	duration := time.Since(startTime)
	
	return &UploadResult{
		URL:      uploader.Url(),
		Size:     dataSize,
		Metadata: metadata,
		Checksum: checksum,
		Duration: duration,
	}, nil
}

// ResumeUpload resumes an existing upload
func (c *TUSClient) ResumeUpload(ctx context.Context, uploadURL, filePath string, options UploadOptions) (*UploadResult, error) {
	startTime := time.Now()
	
	// Open file
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()
	
	// Get file info
	fileInfo, err := file.Stat()
	if err != nil {
		return nil, fmt.Errorf("failed to get file info: %w", err)
	}
	
	fileSize := fileInfo.Size()
	
	// Parse upload URL
	parsedURL, err := url.Parse(uploadURL)
	if err != nil {
		return nil, fmt.Errorf("invalid upload URL: %w", err)
	}
	
	// Create uploader for existing upload
	uploader := &tus.Uploader{
		Client: c.client,
		URL:    parsedURL,
		Upload: &tus.Upload{
			Stream: file,
			Size:   fileSize,
		},
	}
	
	// Configure uploader
	if options.ChunkSize > 0 {
		uploader.ChunkSize = options.ChunkSize
	} else {
		uploader.ChunkSize = c.chunkSize
	}
	
	// Setup progress tracking
	if options.ProgressCallback != nil {
		progress := NewUploadProgress(fileSize, options.ProgressCallback)
		uploader.NotifyUploadProgress = func(bytesUploaded, bytesTotal int64) {
			progress.Update(bytesUploaded)
		}
	}
	
	// Use context if provided
	if options.Context != nil {
		ctx = options.Context
	}
	if ctx == nil {
		ctx = context.Background()
	}
	
	// Resume upload
	err = uploader.Upload(ctx)
	if err != nil {
		return nil, fmt.Errorf("resume upload failed: %w", err)
	}
	
	duration := time.Since(startTime)
	
	return &UploadResult{
		URL:      uploadURL,
		Size:     fileSize,
		Metadata: nil, // Would need to fetch from server
		Checksum: "", // Would need to calculate
		Duration: duration,
	}, nil
}

// GetUploadInfo gets information about an existing upload
func (c *TUSClient) GetUploadInfo(ctx context.Context, uploadURL string) (map[string]interface{}, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	
	req, err := http.NewRequestWithContext(ctx, "HEAD", uploadURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	
	req.Header.Set("Tus-Resumable", "1.0.0")
	
	resp, err := c.client.HttpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to get upload info: %w", err)
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		return nil, fmt.Errorf("server returned status %d", resp.StatusCode)
	}
	
	info := make(map[string]interface{})
	
	if offset := resp.Header.Get("Upload-Offset"); offset != "" {
		if val, err := strconv.ParseInt(offset, 10, 64); err == nil {
			info["offset"] = val
		}
	}
	
	if length := resp.Header.Get("Upload-Length"); length != "" {
		if val, err := strconv.ParseInt(length, 10, 64); err == nil {
			info["length"] = val
		}
	}
	
	if metadata := resp.Header.Get("Upload-Metadata"); metadata != "" {
		info["metadata"] = metadata
	}
	
	if expires := resp.Header.Get("Upload-Expires"); expires != "" {
		info["expires"] = expires
	}
	
	// Calculate completion status
	if offset, ok := info["offset"].(int64); ok {
		if length, ok := info["length"].(int64); ok {
			info["complete"] = offset >= length
			if length > 0 {
				info["progress"] = float64(offset) / float64(length) * 100
			}
		}
	}
	
	return info, nil
}

// DeleteUpload deletes an existing upload
func (c *TUSClient) DeleteUpload(ctx context.Context, uploadURL string) error {
	if ctx == nil {
		ctx = context.Background()
	}
	
	req, err := http.NewRequestWithContext(ctx, "DELETE", uploadURL, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	req.Header.Set("Tus-Resumable", "1.0.0")
	
	resp, err := c.client.HttpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to delete upload: %w", err)
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusNotFound {
		return fmt.Errorf("server returned status %d", resp.StatusCode)
	}
	
	return nil
}

// calculateFileChecksum calculates SHA256 checksum of a file
func calculateFileChecksum(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()
	
	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", err
	}
	
	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

// formatBytes formats bytes as human readable string
func formatBytes(bytes int64) string {
	const unit = 1024
	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}
	div, exp := int64(unit), 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}

// formatDuration formats duration as human readable string
func formatDuration(d time.Duration) string {
	if d < time.Second {
		return fmt.Sprintf("%.0fms", float64(d)/float64(time.Millisecond))
	} else if d < time.Minute {
		return fmt.Sprintf("%.1fs", d.Seconds())
	} else {
		return fmt.Sprintf("%.1fm", d.Minutes())
	}
}

// Example usage
func main() {
	fmt.Println("TUS Golang Client Library")
	fmt.Println("This is a library file. See examples/ directory for usage examples.")
}